include/crab/domains/mru_region_domain.hpp

```diff
@@ -0,0 +1,1336 @@
+    crab::CrabStats::count(domain_name() + ".count.project");
+    crab::ScopedCrabStats __st__(domain_name() + ".project");
+
+    if (is_bottom() || is_top()) {
+      return;
+    }
+
+    variable_vector_t sorted_variables(variables.begin(), variables.end());
+    std::sort(sorted_variables.begin(), sorted_variables.end());
+
+    // -- project in the base domain
+    std::vector<base_variable_t> base_vars;
+    base_vars.reserve(variables.size());
+    for (auto const &v : variables) {
+      if (v.get_type().is_region()) {
+        continue;
+      }
+      get_or_insert_gvars(v).add(base_vars);
+    }
+    // project cache_lines domain
+    m_cache_lines.project(base_vars);
+    // project regs domain
+    m_regs.project(base_vars);
+    // project addrs domain
+    m_addrs.project(base_vars);
+    // project flags domain
+    m_flags.project(base_vars);
+
+    // -- update m_var_map and m_rev_var_map
+    std::vector<variable_t> var_map_to_remove;
+    var_map_to_remove.reserve(m_var_map.size());
+    for (auto &kv : m_var_map) {
+      if (!std::binary_search(sorted_variables.begin(), sorted_variables.end(),
+                              kv.first)) {
+        var_map_to_remove.push_back(kv.first);
+        kv.second.remove_rev_varmap(m_rev_var_map);
+      }
+    }
+    for (auto &v : var_map_to_remove) {
+      m_var_map.erase(v);
+    }
+  }
+
+  std::string domain_name() const override {
+    return "CacheDomain(" + m_cache_lines.domain_name() + ", " +
+           m_regs.domain_name() + ")";
+  }
+
+  void write(crab_os &o) const override {
+    if (is_bottom()) {
+      o << "_|_";
+    } else if (is_top()) {
+      o << "{}";
+    } else {
+      CRAB_LOG("mru-region-print", o << "(CacheLine=" << m_cache_lines << ",\n"
+                                     << "Regs=" << m_regs << ",\n"
+                                     << "Addrs=" << m_addrs << ",\n"
+                                     << "Flags=" << m_flags << ")\n";);
+
+      std::unordered_map<std::string, std::string> renaming_map;
+      auto get_type_fn = [](const variable_t &v) -> variable_type {
+        return variable_type(variable_type_kind::UNK_TYPE);
+      };
+      // Assigns a string name to each ghost variable
+      ghost_variables_t::mk_renaming_map(m_rev_var_map, get_type_fn,
+                                         renaming_map);
+      m_alloc.add_renaming_map(renaming_map);
+      o << "(CacheLine=" << m_cache_lines << ",\n"
+        << "Regs=" << m_regs << ",\n"
+        << "Addrs=" << m_addrs << ",\n"
+        << "Flags=" << m_flags << ")\n";
+      m_alloc.clear_renaming_map();
+    }
+  }
+  /**------------- End domain API definitions -------------------**/
+
+  /**------------- Begin MRU API definitions -------------------**/
+  // perform a similiar work for cache.lines.constraint(cache.regs)
+  // This one requires to be renaming because mem base dom used different base
+  // vars
+  mru_region_domain_t cstrs() {
```

Comment:
it should be private.

Reply:
This seems to be part of a reduction step. We pass all the equalities from `m_regs` and add them to `m_cache_lines`. More comments would be helpful

Reply:
Agree. This one follows the old implementation, will change that in the next iteration.